Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java

```diff
@@ -163,58 +197,75 @@ public static CompactSampler uniformSampler(int sampleSize, long seed, boolean s
+        } else if (weight < this.weight[0]) {
+            long evictedIndex = storeSequenceIndexesEnabled ? this.sequenceIndex[0] : 0L;
+            evictedPoint = new Sequential<>(this.pointIndex[0], this.weight[0], evictedIndex);
             --size;
-            weightArray[0] = weightArray[size];
-            referenceArray[0] = referenceArray[size];
+            this.weight[0] = this.weight[size];
+            this.pointIndex[0] = this.pointIndex[size];
             swapDown(0);
             return Optional.of(weight);
         } else {
             return Optional.empty();
         }
-
     }
 
-    private void swapDown(int startIndex) {
+    /**
+     * Check to see if the weight at current index is greater than or equal to the
+     * weight at each corresponding child index. If validate is true then throw an
+     * IllegalStateException, otherwise swap the nodes and perform the same check at
+     * the next level. Continue until you reach a level where the parent node's
+     * weight is greater than or equal to both children's weights, or until there
+     * are no more levels to descend.
+     *
+     * @param startIndex The index of node to start with.
+     * @param validate   If true, a violation of the heap property will throw an
+     *                   IllegalStateException. If false, then swap nodes that
+     *                   violate the heap property.
+     */
+    private void swapDown(int startIndex, boolean validate) {
         int current = startIndex;
-        while ((2 * current + 1) < size) {
+        while (2 * current + 1 < size) {
             int maxIndex = 2 * current + 1;
-            if ((2 * current + 2 < size) && (weightArray[2 * current + 2] > weightArray[maxIndex]))
+            if (2 * current + 2 < size && weight[2 * current + 2] > weight[maxIndex]) {
                 maxIndex = 2 * current + 2;
-            if (weightArray[maxIndex] > weightArray[current]) {
+            }
+            if (weight[maxIndex] > weight[current]) {
+                if (validate) {
+                    throw new IllegalStateException("the heap property is not satisfied at index " + current);
+                }
                 swapWeights(current, maxIndex);
                 current = maxIndex;
-            } else
+            } else {
                 break;
+            }
         }
     }
 
-    private void reheap() {
+    private void swapDown(int startIndex) {
+        swapDown(startIndex, false);
+    }
+
+    public void reheap(boolean validate) {
         for (int i = (size + 1) / 2; i >= 0; i--) {
-            swapDown(i);
+            swapDown(i, validate);
         }
     }
 
     @Override
-    public void addSample(Integer pointRef, float weight, long seqNUm) {
-        checkState(size < capacity, " sampler full");
-        weightArray[size] = weight;
-        referenceArray[size] = pointRef;
+    public void addSample(Integer pointIndex, float weight, long sequenceIndex) {
```

Comment:
Do we want to check if a acceptSample was called (getEvictedPoint)? And reset that state? That would be true if size = capacity - 1.

Should we check for weights generated earlier in an acceptSample even if nothing was evicted? Probably not. But this does leaf the feature that we can artificially change the weight of a an element that is sampled to have different duration in the sampler.